autonomy/analyse/abci/app_spec.py

```diff
@@ -338,6 +343,40 @@ def abci_to_dfa(cls, abci_app_cls: Any, label: str = "") -> "DFA":
         )
 
 
+event_pattern = re.compile(r"Event\.(\w+)", re.DOTALL)
+
+
+def _check_unreferenced_events(abci_app_cls: AbciApp) -> None:
+    """Checks that events defined in the AbciApp transition function are referenced in the source code of the coresponding round or its superclasses."""
+
+    error_strings = []
+    timeout_events = {k.name for k in abci_app_cls.event_to_timeout.keys()}
+
+    for round_cls, round_transitions in abci_app_cls.transition_function.items():
+        trf_events = {
+            str(e).rsplit(".", 1)[1] for e in round_transitions
+        } - timeout_events
+        referenced_events = set()
+
+        for base in filter(
+            lambda x: x.__class__.__module__ != "builtins",
+            inspect.getmro(round_cls),
+        ):
+            src = textwrap.dedent(inspect.getsource(base))
+            referenced_events.update(event_pattern.findall(src))
+
+        if trf_events.symmetric_difference(referenced_events):
+            error_strings.append(
+                f" - {round_cls.__name__}: transition function events {trf_events} do not match referenced events {referenced_events}."
+            )
+
+    if len(error_strings) > 0:
+        raise DFASpecificationError(
+            f"{abci_app_cls} ABCI App has the following issues:\n"
+            + "\n".join(error_strings)
+        )
```

Comment:
I like this, we return all error in the ABCIApp, not Error out on the first, which would make this a game of iterative and incremental fixes

Reply:
I am glad you like it, as you suggested this approach in an earlier PR :)